packages/lesmis-server/src/routes/PDFExportRoute.ts

```diff
@@ -0,0 +1,91 @@
+/*
+ * Tupaia
+ * Copyright (c) 2017 - 2022 Beyond Essential Systems Pty Ltd
+ *
+ */
+import { Request, Response, NextFunction } from 'express';
+import { Route } from '@tupaia/server-boilerplate';
+import puppeteer from 'puppeteer';
+import Cookies from 'cookies';
+
+type SessionCookies = {
+  sessionCookieName: string;
+  sessionCookieValue: string;
+};
+
+type Body = {
+  pdfPageUrl: string;
+};
+
+export type PDFExportRequest = Request<
+  Record<string, never>,
+  { contents: Buffer },
+  Body,
+  Record<string, any>
+>;
+
+export class PDFExportRoute extends Route<PDFExportRequest> {
+  public constructor(req: PDFExportRequest, res: Response, next: NextFunction) {
+    super(req, res, next);
+    this.type = 'download';
+  }
+
+  private extractSessionCookie = (): SessionCookies => {
+    const cookies = new Cookies(this.req, this.res);
+    const sessionCookieName = 'sessionCookie';
+    const sessionCookieValue = cookies.get(sessionCookieName);
+    if (!sessionCookieValue) {
+      throw new Error(`'sessionCookie' is not found`);
+    }
+    return { sessionCookieName, sessionCookieValue };
+  };
+
+  private verifyBody = (body: any): Body => {
+    const { pdfPageUrl } = body;
+    if (!pdfPageUrl || typeof pdfPageUrl !== 'string') {
+      throw new Error(`'pdfPageUrl' should be provided in request body, got: ${pdfPageUrl}`);
+    }
+
+    return { pdfPageUrl };
+  };
+
+  private exportPDF = async (): Promise<Buffer> => {
+    const { sessionCookieName, sessionCookieValue } = this.extractSessionCookie();
+    const { pdfPageUrl } = this.verifyBody(this.req.body);
+    const location = new URL(pdfPageUrl);
+    let browser;
+    let result;
+
+    try {
+      browser = await puppeteer.launch();
+      const page = await browser.newPage();
+      await page.setCookie({
+        name: sessionCookieName,
+        domain: location.hostname.split(':')[0], // localhost:8030 -> localhost
```

Comment:
Nice thinking using URL. Splitting not necessary anymore, hostname always returns the domain without the port.